// fetch-h2/dist/test/lib/server-common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ignoreError = (cb) => { try {
    cb();
}
catch (err) { } };
class Server {
    constructor() {
        this.port = null;
        this._opts = {};
        this._server = void 0;
    }
    async listen(port = void 0) {
        return new Promise((resolve, _reject) => {
            this._server.listen(port, "0.0.0.0", resolve);
        })
            .then(() => {
            const address = this._server.address();
            if (!address || typeof address === "string")
                return 0;
            return address.port;
        })
            .then(port => {
            this.port = port;
            return port;
        });
    }
    async shutdown() {
        await this._shutdown();
        return new Promise((resolve, reject) => {
            this._server.close((err) => {
                if (err)
                    return reject(err);
                resolve();
            });
        });
    }
    async _shutdown() { }
}
exports.Server = Server;
class TypedServer extends Server {
    constructor() {
        super(...arguments);
        this._server = void 0;
    }
}
exports.TypedServer = TypedServer;
