// fetch-h2/dist/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const url_1 = require("url");
const zlib_1 = require("zlib");
function arrayify(value) {
    if (value != null && Array.isArray(value))
        return value;
    return value == null
        ? []
        : Array.isArray(value)
            ? [...value]
            : [value];
}
exports.arrayify = arrayify;
function parseLocation(location, origin) {
    if ("string" !== typeof location)
        return null;
    const url = new url_1.URL(location, origin);
    return url.href;
}
exports.parseLocation = parseLocation;
exports.isRedirectStatus = {
    300: true,
    301: true,
    302: true,
    303: true,
    305: true,
    307: true,
    308: true,
};
function makeOkError(err) {
    err.metaData = err.metaData || {};
    err.metaData.ok = true;
    return err;
}
exports.makeOkError = makeOkError;
function parseInput(url) {
    const explicitProtocol = (url.startsWith("http2://") || url.startsWith("http1://"))
        ? url.substr(0, 5)
        : null;
    url = url.replace(/^http[12]:\/\//, "http://");
    const { origin, hostname, port, protocol } = new url_1.URL(url);
    return {
        hostname,
        origin,
        port: port || (protocol === "https:" ? "443" : "80"),
        protocol: explicitProtocol || protocol.replace(":", ""),
        url,
    };
}
exports.parseInput = parseInput;
exports.identity = (t) => t;
function uniq(arr, pred) {
    if (!pred)
        return Array.from(new Set(arr));
    const known = new Set();
    return arr.filter(value => {
        const u = pred(value);
        const first = !known.has(u);
        known.add(u);
        return first;
    });
}
exports.uniq = uniq;
function hasBuiltinBrotli() {
    return typeof zlib_1.createBrotliCompress === "function";
}
exports.hasBuiltinBrotli = hasBuiltinBrotli;
